### ISSUE TITLE
Unable to interrupt stream that has been translated

### ISSUE THREAD

**Issue by mpilquist** [typelevel/fs2#2143]
This is the root cause of https://github.com/http4s/http4s/issues/3878.

Reproduction for unit test suite:

```scala
  test("translateInterruptible") {
    type Eff[A] = cats.data.EitherT[IO, String, A]
    val Eff = Async[Eff]
    Stream
      .eval(Eff.never)
      .merge(Stream.eval(Eff.delay(1)).delayBy(5.millis).repeat)
      .interruptAfter(10.millis)
      .translateInterruptible(new (Eff ~> IO) {
        def apply[X](eff: Eff[X]) = eff.value.flatMap {
          case Left(t) => IO.raiseError(new RuntimeException(t))
          case Right(x) => IO.pure(x)
        }
      })
      .compile
      .drain
  }
```

There's a hidden `ClassCastException` occurring here, as a result of a soundness issue with `Stream.getScope`. Consider:

```scala
val s = Stream.getScope[OptionT[IO, *]].
  evalMap(_.interrupt(Right(()))).
  translateInterruptible(new (OptionT[IO, *] ~> IO) {
    def apply[X](o: OptionT[IO, X]) = 
      o.value.flatMap {
        case None => IO.raiseError(new NoSuchElementException)
        case Some(x) => IO.pure(x)
      }
    }
  )

s.compiler.lastOrError.unsafeRunSync
java.lang.ClassCastException: class cats.effect.IO$RaiseError cannot be cast to class cats.data.OptionT (cats.effect.IO$RaiseError and cats.data.OptionT are in unnamed module of loader ammonite.runtime.SpecialClassLoader @71def8f8)
```

The problem is that `getScope` returns the current `CompileScope` that's being used during evaluation, but that's a `CompileScope[IO]`, not a `CompileScope[OptionT[IO, *]]`. The call to `scope.interrupt` returns an `IO` which is then casted to an `OptionT[IO, *]`.

**Comment by mpilquist** [comment]
One potential avenue for fixing is (1) deprecating `getScope` (2) for each operation on `Scope`, provide a new operation on `Stream` companion that returns a `Stream[F, X]` (3) implement those new operations via new nodes in stream algebra.

**Comment by mpilquist** [comment]
On second look, this isn't as straightforward as I had hoped, as going down the path I suggested requires implementing `interruptWhen` in terms of `concurrently` whereas `concurrently` is implemented in terms of `interruptWhen`.

**Comment by mpilquist** [comment]
Fixed in #2145 for develop branch. Will backport a fix to main branch tonight.
